plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -76,13 +84,115 @@ protected function initialize() {
+	 * - sourceData: The context data.
+	 *
+	 * The sourceData structure depends on the context type as follows:
+	 * - site: []
+	 * - order: ['orderId' => int]
+	 * - cart: ['productIds' => array]
+	 * - archive: ['taxonomy' => string, 'termId' => int]
+	 * - product: ['productId' => int]
+	 *
+	 * @param array    $context      The block context.
+	 * @param array    $block        The parsed block.
+	 * @param WP_Block $parent_block The parent block.
+	 *
+	 * @return array The block's context including the location context.
+	 */
+	public function update_context( $context, $block, $parent_block ) {
+		// Run only on frontend.
+		// This is needed to avoid SSR renders within patterns. @see https://github.com/woocommerce/woocommerce/issues/45181.
+		if ( is_admin() || \WC()->is_rest_api_request() ) {
+			return $context;
+		}
+
+		if ( 'woocommerce/product-collection' === $block['blockName']
+			&& isset( $block['attrs']['query']['isProductCollectionBlock'] ) ) {
+					$this->product_collection_inner_blocks_names = array_reverse(
+						$this->extract_inner_block_names( $block )
+					);
+		}
+
+		$this->provide_location_context_inner_blocks( $block, $context );
+
+		return $context;
+	}
+
+	/**
+	 * Extract the inner block names for the Product Collection block.
+	 *
+	 * @param array $block The Product Collection block.
+	 * @param array $result Array of inner block names.
+	 *
+	 * @return array Array containing all the inner block names.
+	 */
+	protected function extract_inner_block_names( $block, &$result = array() ) {
+		if ( isset( $block['blockName'] ) ) {
+			$result[] = $block['blockName'];
+		}
+
+		if ( isset( $block['innerBlocks'] ) ) {
+			foreach ( $block['innerBlocks'] as $inner_block ) {
+				$this->extract_inner_block_names( $inner_block, $result );
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Add the productCollectionLocation context data into the inner blocks' context.
+	 *
+	 * @param array $block Block attributes.
+	 * @param array $context Block context.
+	 */
+	protected function provide_location_context_inner_blocks( $block, &$context ) {
+		if ( empty( $this->product_collection_inner_blocks_names ) ) {
+			return;
+		}
+
+		$block_name           = array_pop( $this->product_collection_inner_blocks_names );
+		$is_in_single_product = isset( $context['singleProduct'] ) && ! empty( $context['postId'] );
+
+		if ( $block_name === $block['blockName'] ) {
+
+			$context['productCollectionLocation'] = $is_in_single_product ? array(
+				'type'       => 'product',
+				'sourceData' => array(
+					'productId' => absint( $context['postId'] ),
+				),
+			) : $this->get_location_context();
+		}
+	}
+
+	/**
+	 * Get the global location context.
+	 * Serve as a runtime cache for the location context.
+	 *
+	 * @return array The location context as described in self::update_context().
+	 */
+	protected function get_location_context() {
+		static $location_context = null;
```

Comment:
Interesting! I noticed the $location_context variable is marked as static here. Am I correct in assuming this is for caching the location context across multiple calls to this function within the same request lifecycle, to optimize performance by avoiding redundant computations? Could you please confirm if this is the primary reason, or if there are other considerations for choosing to use a static variable in this context?

Reply:
Exactly! By declaring this as _static_, we can avoid repeatedly parsing the global context while looping through the collection's inner blocks.